File: csharp/AStarPathfinder.cs
using System;
using System.Collections.Generic;

public class AStarPathfinder : Benchmark
{
    private class Node : IComparable<Node>
    {
        public int X { get; }
        public int Y { get; }
        public int FScore { get; }

        public Node(int x, int y, int fScore)
        {
            X = x;
            Y = y;
            FScore = fScore;
        }

        public int CompareTo(Node? other)
        {
            if (other == null) return 1;

            if (FScore != other.FScore) return FScore.CompareTo(other.FScore);
            if (Y != other.Y) return Y.CompareTo(other.Y);
            return X.CompareTo(other.X);
        }
    }

    private class BinaryHeap
    {
        private readonly List<Node> _data;

        public BinaryHeap(int initialCapacity = 64)
        {
            _data = new List<Node>(initialCapacity);
        }

        public void Push(Node item)
        {
            _data.Add(item);
            SiftUp(_data.Count - 1);
        }

        public Node? Pop()
        {
            if (_data.Count == 0) return null;

            if (_data.Count == 1)
            {
                Node result = _data[0];
                _data.RemoveAt(0);
                return result;
            }

            Node resultNode = _data[0];
            _data[0] = _data[_data.Count - 1];
            _data.RemoveAt(_data.Count - 1);
            SiftDown(0);
            return resultNode;
        }

        public bool IsEmpty() => _data.Count == 0;

        private void SiftUp(int index)
        {
            while (index > 0)
            {
                int parent = (index - 1) >> 1;
                if (_data[index].CompareTo(_data[parent]) >= 0) break;
                Swap(index, parent);
                index = parent;
            }
        }

        private void SiftDown(int index)
        {
            int size = _data.Count;
            while (true)
            {
                int left = (index << 1) + 1;
                int right = left + 1;
                int smallest = index;

                if (left < size && _data[left].CompareTo(_data[smallest]) < 0)
                    smallest = left;
                if (right < size && _data[right].CompareTo(_data[smallest]) < 0)
                    smallest = right;

                if (smallest == index) break;

                Swap(index, smallest);
                index = smallest;
            }
        }

        private void Swap(int i, int j)
        {
            Node temp = _data[i];
            _data[i] = _data[j];
            _data[j] = temp;
        }
    }

    private uint _result;
    private int _startX;
    private int _startY;
    private int _goalX;
    private int _goalY;
    private int _width;
    private int _height;
    private bool[,] _mazeGrid;

    private int[] _gScoresCache;
    private int[] _cameFromCache;

    private static readonly (int dx, int dy)[] _directions = new[]
    {
        (0, -1), (1, 0), (0, 1), (-1, 0)
    };

    private const int STRAIGHT_COST = 1000;

    public AStarPathfinder()
    {
        _width = (int)ConfigVal("w");
        _height = (int)ConfigVal("h");
        _startX = 1;
        _startY = 1;
        _goalX = _width - 2;
        _goalY = _height - 2;

        int size = _width * _height;
        _gScoresCache = new int[size];
        _cameFromCache = new int[size];
    }

    private bool[,] GenerateWalkableMaze(int width, int height)
    {
        return MazeGenerator.Maze.GenerateWalkableMaze(width, height);
    }

    private int Distance(int aX, int aY, int bX, int bY)
    {
        return Math.Abs(aX - bX) + Math.Abs(aY - bY);
    }

    private int PackCoords(int x, int y) => y * _width + x;

    private (int x, int y) UnpackCoords(int packed) => (packed % _width, packed / _width);

    private (List<(int x, int y)>? path, int nodesExplored) FindPath()
    {
        bool[,] grid = _mazeGrid;
        int width = _width;
        int height = _height;

        int[] gScores = _gScoresCache;
        int[] cameFrom = _cameFromCache;

        Array.Fill(gScores, int.MaxValue);
        Array.Fill(cameFrom, -1);

        BinaryHeap openSet = new BinaryHeap(width * height);
        int nodesExplored = 0;

        int startIdx = PackCoords(_startX, _startY);
        gScores[startIdx] = 0;
        openSet.Push(new Node(_startX, _startY, Distance(_startX, _startY, _goalX, _goalY)));

        while (!openSet.IsEmpty())
        {
            Node? current = openSet.Pop();
            if (current == null) break;

            nodesExplored++;

            if (current.X == _goalX && current.Y == _goalY)
            {
                List<(int x, int y)> path = new List<(int, int)>();
                int x = current.X;
                int y = current.Y;

                while (x != _startX || y != _startY)
                {
                    path.Add((x, y));
                    int idx = PackCoords(x, y);
                    int packed = cameFrom[idx];
                    if (packed == -1) break;

                    var (prevX, prevY) = UnpackCoords(packed);
                    x = prevX;
                    y = prevY;
                }

                path.Add((_startX, _startY));
                path.Reverse();
                return (path, nodesExplored);
            }

            int currentIdx = PackCoords(current.X, current.Y);
            int currentG = gScores[currentIdx];

            foreach (var (dx, dy) in _directions)
            {
                int nx = current.X + dx;
                int ny = current.Y + dy;

                if (nx < 0 || nx >= width || ny < 0 || ny >= height) continue;
                if (!grid[ny, nx]) continue;

                int tentativeG = currentG + STRAIGHT_COST;
                int neighborIdx = PackCoords(nx, ny);

                if (tentativeG < gScores[neighborIdx])
                {

                    cameFrom[neighborIdx] = currentIdx;
                    gScores[neighborIdx] = tentativeG;

                    int fScore = tentativeG + Distance(nx, ny, _goalX, _goalY);
                    openSet.Push(new Node(nx, ny, fScore));
                }
            }
        }

        return (null, nodesExplored);
    }

    public override void Prepare()
    {
        _mazeGrid = GenerateWalkableMaze(_width, _height);
    }

    public override void Run(long IterationId)
    {
        var (path, nodesExplored) = FindPath();

        uint localResult = 0;

        localResult = (uint)(path?.Count ?? 0);

        localResult = (localResult << 5) + (uint)nodesExplored;

        _result = (uint)((long)_result + localResult);
    }

    public override uint Checksum => _result;
}

File: fsharp/AStarPathfinder.fs
namespace Benchmarks

open System
open System.Collections.Generic

[<Struct; CustomComparison; CustomEquality>]
type ANode =
    val X: int
    val Y: int
    val FScore: int

    new(x, y, fScore) = { X = x; Y = y; FScore = fScore }

    interface IComparable with
        member this.CompareTo(other: obj) =
            if other :? ANode then
                let otherANode = other :?> ANode
                if this.FScore <> otherANode.FScore then
                    this.FScore.CompareTo(otherANode.FScore)
                elif this.Y <> otherANode.Y then
                    this.Y.CompareTo(otherANode.Y)
                else
                    this.X.CompareTo(otherANode.X)
            else
                1

    interface IComparable<ANode> with
        member this.CompareTo(other) =
            if this.FScore <> other.FScore then
                this.FScore.CompareTo(other.FScore)
            elif this.Y <> other.Y then
                this.Y.CompareTo(other.Y)
            else
                this.X.CompareTo(other.X)

    override this.Equals(other: obj) =
        match other with
        | :? ANode as otherANode ->
            this.X = otherANode.X && this.Y = otherANode.Y && this.FScore = otherANode.FScore
        | _ -> false

    override this.GetHashCode() =
        HashCode.Combine(this.X, this.Y, this.FScore)

    interface IEquatable<ANode> with
        member this.Equals(other) =
            this.X = other.X && this.Y = other.Y && this.FScore = other.FScore

type BinaryHeap(initialCapacity: int) =
    let mutable data = Array.zeroCreate<ANode> initialCapacity
    let mutable count = 0

    let ensureCapacity () =
        if count >= data.Length then
            Array.Resize(&data, data.Length * 2)

    let swap i j =
        let temp = data.[i]
        data.[i] <- data.[j]
        data.[j] <- temp

    let siftUp index =
        let mutable idx = index
        let nodes = data

        while idx > 0 do
            let parent = (idx - 1) >>> 1
            if nodes.[idx].FScore >= nodes.[parent].FScore then
                idx <- 0  
            else
                swap idx parent
                idx <- parent

    let siftDown index =
        let mutable idx = index
        let nodes = data
        let size = count
        let mutable continueLoop = true

        while continueLoop do
            let left = (idx <<< 1) + 1
            let right = left + 1
            let mutable smallest = idx

            if left < size && nodes.[left].FScore < nodes.[smallest].FScore then
                smallest <- left

            if right < size && nodes.[right].FScore < nodes.[smallest].FScore then
                smallest <- right

            if smallest = idx then
                continueLoop <- false
            else
                swap idx smallest
                idx <- smallest

    member this.Push(item: ANode) =
        ensureCapacity()
        data.[count] <- item
        count <- count + 1
        siftUp(count - 1)

    member this.Pop() =
        if count = 0 then None
        else
            let result = data.[0]
            count <- count - 1

            if count > 0 then
                data.[0] <- data.[count]
                siftDown(0)

            Some result

    member _.IsEmpty() = count = 0

type AStarPathfinder() =
    inherit Benchmark()

    [<Literal>]
    let STRAIGHT_COST = 1000

    let directions = [| struct (0, -1); struct (1, 0); struct (0, 1); struct (-1, 0) |]

    let mutable width = 0
    let mutable height = 0
    let mutable startX = 1
    let mutable startY = 1
    let mutable goalX = 0
    let mutable goalY = 0
    let mutable mazeGrid = Array2D.zeroCreate<bool> 0 0
    let mutable gScoresCache = Array.empty<int>
    let mutable cameFromCache = Array.empty<int>
    let mutable result = 0u

    let distance aX aY bX bY =
        abs (aX - bX) + abs (aY - bY)

    let packCoords x y width = y * width + x

    let unpackCoords packed width = 
        (packed % width, packed / width)

    let fillArray (arr: 'T[]) (value: 'T) =
        for i = 0 to arr.Length - 1 do
            arr.[i] <- value

    let findPath () =
        let grid = mazeGrid
        let w = width
        let h = height
        let gScores = gScoresCache
        let cameFrom = cameFromCache

        fillArray gScores Int32.MaxValue
        fillArray cameFrom -1

        let openSet = BinaryHeap(w * h)
        let mutable nodesExplored = 0

        let startIdx = packCoords startX startY w
        gScores.[startIdx] <- 0
        openSet.Push(ANode(startX, startY, distance startX startY goalX goalY))

        let mutable foundPath: (int * int) list option = None

        while not (openSet.IsEmpty()) && foundPath.IsNone do
            match openSet.Pop() with
            | Some current ->
                nodesExplored <- nodesExplored + 1

                if current.X = goalX && current.Y = goalY then
                    let path = ResizeArray<int * int>()
                    let mutable x = current.X
                    let mutable y = current.Y

                    while x <> startX || y <> startY do
                        path.Add((x, y))
                        let idx = packCoords x y w
                        let packed = cameFrom.[idx]

                        if packed = -1 then
                            x <- startX
                            y <- startY
                        else
                            let (prevX, prevY) = unpackCoords packed w
                            x <- prevX
                            y <- prevY

                    path.Add((startX, startY))
                    let reversed = List.ofSeq path
                    foundPath <- Some (List.rev reversed)
                else
                    let currentIdx = packCoords current.X current.Y w
                    let currentG = gScores.[currentIdx]

                    for i = 0 to directions.Length - 1 do
                        let struct (dx, dy) = directions.[i]
                        let nx = current.X + dx
                        let ny = current.Y + dy

                        if nx >= 0 && nx < w && ny >= 0 && ny < h && grid.[ny, nx] then
                            let tentativeG = currentG + STRAIGHT_COST
                            let neighborIdx = packCoords nx ny w

                            if tentativeG < gScores.[neighborIdx] then
                                cameFrom.[neighborIdx] <- currentIdx
                                gScores.[neighborIdx] <- tentativeG

                                let fScore = tentativeG + distance nx ny goalX goalY
                                openSet.Push(ANode(nx, ny, fScore))
            | None -> ()

        match foundPath with
        | Some path -> (Some path, nodesExplored)
        | None -> (None, nodesExplored)

    override this.Checksum = result

    override this.Prepare() =
        width <- int (this.ConfigVal("w"))
        height <- int (this.ConfigVal("h"))
        startX <- 1
        startY <- 1
        goalX <- width - 2
        goalY <- height - 2

        let size = width * height
        gScoresCache <- Array.zeroCreate size
        cameFromCache <- Array.zeroCreate size

        mazeGrid <- MazeTypes.Maze.GenerateWalkableMaze(width, height)
        result <- 0u

    override this.Run(_: int64) =
        let (path, nodesExplored) = findPath()

        let mutable localResult = 0u
        localResult <- uint32 (match path with Some p -> p.Length | None -> 0)
        localResult <- (localResult <<< 5) + uint32 nodesExplored
        result <- result + localResult